pkg/util/deadlock-detector.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util"
+
+	"github.com/golang/glog"
+)
+
+type deadlockDetector struct {
+	name  string
+	lock  sync.RWMutex
+	clock util.Clock
+}
+
+// DeadlockWatchdog creates a watchdog on mutex.  If the mutex can not be acquired
+// within 'maxLockPeriod', the program exits via glog.Exitf() or os.Exit() if that fails
+// 'name' is a semantic name that is useful for the user and is printed on exit.
+func DeadlockWatchdog(lock sync.RWMutex, name string, maxLockPeriod time.Duration) {
```

Comment:
lock needs to be a pointer. Also why not take the interface, that way it will work with regular mutexes, too.